File: frontend/src/lib/enrolment.ts
import { z } from 'zod'
import { api } from './api'

/** Portal configuration — the UI renders whatever the backend config declares. */
export const requirementSchema = z.object({
  accepted_formats: z.array(z.string()),
  max_bytes: z.number(),
  min_bytes: z.number(),
  min_width: z.number().nullable(),
  min_height: z.number().nullable(),
  max_width: z.number().nullable(),
  max_height: z.number().nullable(),
  min_dpi: z.number().nullable(),
  max_pages: z.number().nullable(),
  colour_mode: z.string(),
})

export const documentTypeSchema = z.object({
  id: z.string(),
  label: z.string(),
  short_label: z.string(),
  help: z.string(),
  examples: z.array(z.string()),
  requirement: requirementSchema,
})

export const applicantTypeSchema = z.object({
  id: z.string(),
  label: z.string(),
  description: z.string(),
  required_documents: z.array(z.string()),
  is_primary_demo: z.boolean(),
})

export const portalSchema = z.object({
  portal_id: z.string(),
  portal_name: z.string(),
  authority_note: z.string(),
  journey_note: z.string(),
  config_version: z.string(),
  applicant_types: z.array(applicantTypeSchema),
})

export type Requirement = z.infer<typeof requirementSchema>
export type DocumentType = z.infer<typeof documentTypeSchema>
export type ApplicantType = z.infer<typeof applicantTypeSchema>
export type Portal = z.infer<typeof portalSchema>

export const AADHAAR = 'aadhaar'

export async function fetchPortal(portalId = AADHAAR): Promise<Portal> {
  const { data } = await api.get(`/portals/${portalId}`)
  return portalSchema.parse(data)
}

export async function fetchDocumentTypes(
  applicantType: string,
  portalId = AADHAAR,
): Promise<DocumentType[]> {
  const { data } = await api.get(`/portals/${portalId}/documents`, {
    params: { applicant_type: applicantType },
  })
  return z.array(documentTypeSchema).parse(data)
}

/** Enrolments */
export const personalDetailsSchema = z.object({
  full_name: z.string().trim().min(2, 'Enter the full name'),
  date_of_birth: z.string().min(1, 'Enter the date of birth'),
  gender: z.enum(['male', 'female', 'transgender']),
  guardian_name: z.string().trim().max(120).optional().nullable(),
  email: z.string().trim().email('Enter a valid email').optional().nullable(),
  mobile: z
    .string()
    .trim()
    .regex(/^[6-9]\d{9}$/, 'Enter a valid 10-digit mobile number')
    .optional()
    .nullable(),
})

export const addressSchema = z.object({
  address_line1: z.string().trim().min(3, 'Enter the house / street details'),
  address_line2: z.string().trim().max(160).optional().nullable(),
  landmark: z.string().trim().max(120).optional().nullable(),
  village_town_city: z.string().trim().min(2, 'Enter the village, town or city'),
  district: z.string().trim().min(2, 'Enter the district'),
  state: z.string().trim().min(2, 'Enter the state'),
  pincode: z
    .string()
    .trim()
    .regex(/^[1-9]\d{5}$/, 'Enter a valid 6-digit PIN code'),
})

export const progressSchema = z.object({
  applicant_type: z.boolean(),
  personal_details: z.boolean(),
  address: z.boolean(),
  documents: z.boolean(),
  documents_required: z.array(z.string()).default([]),
  documents_accepted: z.array(z.string()).default([]),
  can_prepare: z.boolean(),
})

export const enrolmentSchema = z.object({
  id: z.number(),
  portal_id: z.string(),
  applicant_type: z.string(),
  status: z.string(),
  personal_details: personalDetailsSchema.nullable().optional(),
  address: addressSchema.nullable().optional(),
  reference_code: z.string().nullable().optional(),
  prepared_at: z.string().nullable().optional(),
  created_at: z.string(),
  updated_at: z.string(),
})

export const enrolmentDetailSchema = enrolmentSchema.extend({ progress: progressSchema })

export type PersonalDetails = z.infer<typeof personalDetailsSchema>
export type AddressInput = z.infer<typeof addressSchema>
export type Enrolment = z.infer<typeof enrolmentSchema>
export type EnrolmentDetail = z.infer<typeof enrolmentDetailSchema>

export async function createEnrolment(applicantType: string): Promise<EnrolmentDetail> {
  const { data } = await api.post('/enrolments', {
    applicant_type: applicantType,
    portal_id: AADHAAR,
  })
  return enrolmentDetailSchema.parse(data)
}

export async function fetchEnrolments(): Promise<Enrolment[]> {
  const { data } = await api.get('/enrolments')
  return z.array(enrolmentSchema).parse(data)
}

export async function fetchEnrolment(id: number): Promise<EnrolmentDetail> {
  const { data } = await api.get(`/enrolments/${id}`)
  return enrolmentDetailSchema.parse(data)
}

export async function updateEnrolment(
  id: number,
  patch: {
    applicant_type?: string
    personal_details?: PersonalDetails
    address?: AddressInput
  },
): Promise<EnrolmentDetail> {
  const { data } = await api.patch(`/enrolments/${id}`, patch)
  return enrolmentDetailSchema.parse(data)
}

export async function prepareEnrolment(id: number): Promise<EnrolmentDetail> {
  const { data } = await api.post(`/enrolments/${id}/prepare`)
  return enrolmentDetailSchema.parse(data)
}

export async function deleteEnrolment(id: number): Promise<void> {
  await api.delete(`/enrolments/${id}`)
}

/** Human-readable byte size, e.g. 2097152 -> "2 MB". */
export function formatBytes(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`
  const kb = bytes / 1024
  if (kb < 1024) return `${Math.round(kb)} KB`
  const mb = kb / 1024
  return `${mb % 1 === 0 ? mb : mb.toFixed(2)} MB`
}

File: backend/app/schemas/enrolment.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

Gender = Literal["male", "female", "transgender"]


class PersonalDetails(BaseModel):
    """Synthetic demo details. No Aadhaar number is collected or stored."""

    full_name: str = Field(min_length=2, max_length=120)
    date_of_birth: date
    gender: Gender
    guardian_name: str | None = Field(default=None, max_length=120)
    email: str | None = Field(default=None, max_length=255)
    mobile: str | None = Field(default=None, max_length=15)

    @field_validator("date_of_birth")
    @classmethod
    def _not_future(cls, v: date) -> date:
        if v > date.today():
            raise ValueError("Date of birth cannot be in the future")
        if v.year < 1900:
            raise ValueError("Enter a valid date of birth")
        return v

    @field_validator("full_name", "guardian_name")
    @classmethod
    def _tidy(cls, v: str | None) -> str | None:
        return " ".join(v.split()) if v else v


class Address(BaseModel):
    address_line1: str = Field(min_length=3, max_length=160)
    address_line2: str | None = Field(default=None, max_length=160)
    landmark: str | None = Field(default=None, max_length=120)
    village_town_city: str = Field(min_length=2, max_length=80)
    district: str = Field(min_length=2, max_length=80)
    state: str = Field(min_length=2, max_length=80)
    pincode: str = Field(pattern=r"^[1-9]\d{5}$")


class EnrolmentCreate(BaseModel):
    applicant_type: str = Field(min_length=2, max_length=40)
    portal_id: str = Field(default="aadhaar", max_length=40)


class EnrolmentUpdate(BaseModel):
    """Partial update — the wizard saves one step at a time."""

    applicant_type: str | None = Field(default=None, max_length=40)
    personal_details: PersonalDetails | None = None
    address: Address | None = None


class EnrolmentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    portal_id: str
    applicant_type: str
    status: str
    personal_details: PersonalDetails | None = None
    address: Address | None = None
    reference_code: str | None = None
    prepared_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class EnrolmentProgress(BaseModel):
    """Which wizard steps are done — drives the UI without hardcoding rules in React."""

    applicant_type: bool
    personal_details: bool
    address: bool
    documents: bool
    documents_required: list[str] = []
    documents_accepted: list[str] = []
    can_prepare: bool


class EnrolmentDetail(EnrolmentOut):
    progress: EnrolmentProgress
